rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -0,0 +1,166 @@
+    if (old_action.sa_sigaction != NULL) {
+      old_action.sa_sigaction(signal_value, siginfo, context);
+    }
+  } else {
+    // *INDENT-OFF*
+    if (
+      old_action.sa_handler != NULL && // Is set
+      old_action.sa_handler != SIG_DFL && // Is not default
+      old_action.sa_handler != SIG_IGN) // Is not ignored
+    // *INDENT-ON*
+    {
+      old_action.sa_handler(signal_value);
+    }
+  }
+#else
+  if (old_signal_handler) {
+    old_signal_handler(signal_value);
+  }
+#endif
+  g_signal_status = signal_value;
+  rmw_ret_t status = rmw_trigger_guard_condition(g_sigint_guard_cond_handle);
+  if (status != RMW_RET_OK) {
+    fprintf(stderr,
+      "[rclcpp::error] failed to trigger guard condition: %s\n", rmw_get_error_string_safe());
+  }
+  g_interrupt_condition_variable.notify_all();
+}
+
+namespace rclcpp
+{
+
+namespace utilities
+{
+
+void
+init(int argc, char * argv[])
+{
+  (void)argc;
+  (void)argv;
+  rmw_ret_t status = rmw_init();
+  if (status != RMW_RET_OK) {
+    // *INDENT-OFF* (prevent uncrustify from making unecessary indents here)
+    throw std::runtime_error(
+      std::string("failed to initialize rmw implementation: ") + rmw_get_error_string_safe());
+    // *INDENT-ON*
+  }
+#ifdef HAS_SIGACTION
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  sigemptyset(&action.sa_mask);
+  action.sa_sigaction = ::signal_handler;
+  action.sa_flags = SA_SIGINFO;
+  ssize_t ret = sigaction(SIGINT, &action, &old_action);
+  if (ret == -1)
+#else
+  ::old_signal_handler = std::signal(SIGINT, ::signal_handler);
+  if (::old_signal_handler == SIG_ERR)
+#endif
+  {
+    const size_t error_length = 1024;
+    char error_string[error_length];
+#ifndef _WIN32
+    strerror_r(errno, error_string, error_length);
+#else
+    strerror_s(error_string, error_length, errno);
+#endif
+    // *INDENT-OFF*
+    throw std::runtime_error(
+      std::string("Failed to set SIGINT signal handler: (" + std::to_string(errno) + ")") +
+      error_string);
+    // *INDENT-ON*
+  }
+}
+
+bool
+ok()
+{
+  return ::g_signal_status == 0;
+}
+
+void
+shutdown()
+{
+  g_signal_status = SIGINT;
+  rmw_ret_t status = rmw_trigger_guard_condition(g_sigint_guard_cond_handle);
+  if (status != RMW_RET_OK) {
+    fprintf(stderr,
+      "[rclcpp::error] failed to trigger guard condition: %s\n", rmw_get_error_string_safe());
+  }
+  g_interrupt_condition_variable.notify_all();
+}
+
+rmw_guard_condition_t *
+get_global_sigint_guard_condition()
+{
+  return ::g_sigint_guard_cond_handle;
+}
+
+bool
+sleep_for(const std::chrono::nanoseconds & nanoseconds)
+{
+  // TODO: determine if posix's nanosleep(2) is more efficient here
+  std::unique_lock<std::mutex> lock(::g_interrupt_mutex);
+  auto cvs = ::g_interrupt_condition_variable.wait_for(lock, nanoseconds);
+  // Return true if the timeout elapsed successfully, otherwise false.
+  return cvs != std::cv_status::no_timeout;
+}
+
+} /* namespace utilities */
+} /* namespace rclcpp */
```

Comment:
I'm going to do a follow up pr to make it cpplint compliant. I'll do that at the same time as rcl and rclc. A more salient point is that I shouldn't use the namespace bracket like that in the cpp file and this file is inconsistent with the other cpp file's style. Fixed in 0b59eba.